Nightly cuda/11.2 test failures with cusparse enabled: sparse_cuda, graph_cuda, wiki_spgemm

Test failures with the sparse_cuda, graph_cuda, wiki_spgemm began with cuda/11.2 builds on Power9+Volta70 with cusparse and cublas tpls enabled, some likely related to merge of #1618 

**Test failure snips**:
graph_cuda
```
03:41:57 [ RUN      ] cuda.graph_random_graph_coarsen_double_int_int_TestExecSpace
03:41:57 unknown file: Failure
03:41:57 C++ exception with description "cusparseSpGEMM_compute( h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta, h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT, h->spgemmDescr, &h->bufferSize4, h->buffer4) error( CUSPARSE_STATUS_INSUFFICIENT_RESOURCES): insufficient resources /home/jenkins/weaver-new/workspace/KokkosKernels_Weaver_Cuda_Serial_cuda_1122_gcc_831_cusparse_cublas/kokkos-kernels/sparse/tpls/KokkosSparse_spgemm_symbolic_tpl_spec_decl.hpp:275" thrown in the test body.
03:41:57 [  FAILED  ] cuda.graph_random_graph_coarsen_double_int_int_TestExecSpace (1300 ms)
03:41:57 [ RUN      ] cuda.graph_grid_graph_coarsen_double_int_int_TestExecSpace
03:41:57 cudaFree(arg_alloc_ptr) error( cudaErrorIllegalAddress): an illegal memory access was encountered /home/jenkins/weaver-new/workspace/KokkosKernels_Weaver_Cuda_Serial_cuda_1122_gcc_831_cusparse_cublas/kokkos/core/src/Cuda/Kokkos_CudaSpace.cpp:334
03:41:57 Backtrace:
03:41:57                       [0x10874964]
03:41:57                       [0x1086b804]
03:41:57                       [0x1086b8f8]
03:41:57                       [0x1087c7fc]
03:41:57                       [0x10877bc0]
03:41:57                       [0x10877cb0]
03:41:57                       [0x10877d50]
03:41:57                       [0x103e3560]
03:41:57                       [0x103e35b0]
03:41:57                       [0x103e3650]
03:41:57                       [0x10874674]
03:41:57                       [0x1008fde0]
03:41:57                       [0x1008fe34]
03:41:57                       [0x100a8e6c]
03:41:57                       [0x100a99fc]
03:41:57                       [0x10089314]
03:41:57                       [0x10015bac]
03:41:57                       [0x10833d40]
03:41:57                       [0x1082c670]
03:41:57                       [0x108073b0]
03:41:57                       [0x10807f14]
03:41:57                       [0x10808978]
03:41:57                       [0x10811ee0]
03:41:57                       [0x108356ac]
03:41:57                       [0x1082dcc8]
03:41:57                       [0x10810538]
03:41:57                       [0x100134dc]
03:41:57                       [0x100128f8]
03:41:57                   [0x200016684078]
03:41:57 __libc_start_main [0x200016684264]
```

sparse_cuda
```
03:43:56 [ RUN      ] cuda.sparse_spgemm_jacobi_double_int_int_TestExecSpace
03:43:56 cudaStreamSynchronize(stream) error( cudaErrorIllegalAddress): an illegal memory access was encountered /home/jenkins/weaver-new/workspace/KokkosKernels_Weaver_Cuda_Serial_cuda_1122_gcc_831_cusparse_cublas/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:144
03:43:56 Backtrace:
03:43:56                       [0x11d7d1f4]
03:43:56                       [0x11d73ef4]
03:43:56                       [0x11d73fe8]
03:43:56                       [0x11d8508c]
03:43:56                       [0x11d86920]
03:43:56                       [0x11d86b18]
03:43:56                       [0x11503f30]
03:43:56                       [0x11500eac]
03:43:56                       [0x102d2880]
03:43:56                       [0x1017e178]
03:43:56                       [0x100cc8a8]
03:43:56                       [0x100233c4]
03:43:56                       [0x11d3cc74]
03:43:56                       [0x11d35b70]
03:43:56                       [0x11d10c00]
03:43:56                       [0x11d11764]
03:43:56                       [0x11d121c8]
03:43:56                       [0x11d1b730]
03:43:56                       [0x11d3e5e0]
03:43:56                       [0x11d370e0]
03:43:56                       [0x11d19d88]
03:43:56                       [0x1001db5c]
03:43:56                       [0x1001cf78]
03:43:56                   [0x200016684078]
03:43:56 __libc_start_main [0x200016684264]
```

wiki_spgemm
```
03:57:46 18/24 Test #18: wiki_spgemm ......................Subprocess aborted***Exception:   1.24 sec
03:57:46 terminate called after throwing an instance of 'std::runtime_error'
03:57:46   what():  cusparseSpGEMM_compute( h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta, h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT, h->spgemmDescr, &h->bufferSize4, h->buffer4) error( CUSPARSE_STATUS_INTERNAL_ERROR): internal error /home/jenkins/weaver-new/workspace/KokkosKernels_Weaver_Cuda_Serial_cuda_1122_gcc_831_cusparse_cublas/kokkos-kernels/sparse/tpls/KokkosSparse_spgemm_numeric_tpl_spec_decl.hpp:180
```

**Reproducer** (weaver rhel8 queue):
```bash
module load cuda/11.2.2/gcc/8.3.1 cmake/3.23.1

cm_generate_makefile.bash --with-cuda --with-serial --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --arch=Volta70,Power9 --with-cuda-options=enable_lambda --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --with-tpls=cusparse,cublas --cxxstandard=17
```

@brian-kelley are you able to take a look?

@ndellingwood Could you push a Trilinos branch (to your fork or wherever) that has the needed changes to use KokkosKernels with these patches? I tried applying #1618 as a patch to 3.7.1, but that'll be a pain because it's from before the directory structure change.

Then I tried dropping just KokkosKernels develop into Trilinos develop, but got:
```
/ascldap/users/bmkelle/TrilinosWork/Trilinos/packages/tpetra/core/src/Tpetra_LocalCrsMatrixOperator_decl.hpp:45:10: fatal error: KokkosSparse_CrsMatrix.hpp: No such file or directory
   45 | #include "KokkosSparse_CrsMatrix.hpp"
```
so I assume that you had to work through some CMake changes in KokkosKernels to export the new include directories?

--- a/sparse/tpls/KokkosSparse_spgemm_symbolic_tpl_spec_decl.hpp
+++ b/sparse/tpls/KokkosSparse_spgemm_symbolic_tpl_spec_decl.hpp
@@ -206,84 +206,118 @@ void spgemm_symbolic_cusparse(KernelHandle *handle, lno_t m, lno_t n, lno_t k,
                               const ConstEntriesType &entriesA,
                               const ConstRowMapType &row_mapB,
                               const ConstEntriesType &entriesB,
-                              const RowMapType &row_mapC,
-                              bool /* computeRowptrs */) {
-  using scalar_type = typename KernelHandle::nnz_scalar_t;
-  using Offset      = typename KernelHandle::size_type;
-  if (handle->is_symbolic_called() && handle->are_rowptrs_computed()) return;
-  handle->create_cusparse_spgemm_handle(false, false);
-  auto h = handle->get_cusparse_spgemm_handle();
+                              const RowMapType &row_mapC, bool computeRowptrs) {
+  using scalar_type      = typename KernelHandle::nnz_scalar_t;
+  using ordinal_type     = typename KernelHandle::nnz_lno_t;
+  const auto alpha       = Kokkos::ArithTraits<scalar_type>::one();
+  const auto beta        = Kokkos::ArithTraits<scalar_type>::zero();
+  void *dummyValues_AB   = nullptr;
+  bool firstSymbolicCall = false;
+  if (!handle->is_symbolic_called()) {
+    handle->create_cusparse_spgemm_handle(false, false);
+    auto h = handle->get_cusparse_spgemm_handle();
 
-  // Follow
-  // https://github.com/NVIDIA/CUDALibrarySamples/blob/master/cuSPARSE/spgemm
+    // Follow
+    // https://github.com/NVIDIA/CUDALibrarySamples/blob/master/cuSPARSE/spgemm
 
-  const auto alpha = Kokkos::ArithTraits<scalar_type>::one();
-  const auto beta  = Kokkos::ArithTraits<scalar_type>::zero();
+    // In non-reuse interface, forced to give A,B dummy values to
+    // cusparseSpGEMM_compute. And it actually reads them, so they must be
+    // allocated and of the correct type. This compute will be called again in
+    // numeric with the real values.
+    //
+    // The dummy values can be uninitialized. cusparseSpGEMM_compute does
+    // not remove numerical zeros from the sparsity pattern.
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(
+        &dummyValues_AB, sizeof(scalar_type) *
+                             std::max(entriesA.extent(0), entriesB.extent(0))));
 
-  // In non-reuse interface, forced to give A,B dummy values to
-  // cusparseSpGEMM_compute. And it actually reads them, so they must be
-  // allocated and of the correct type. This compute will be called again in
-  // numeric with the real values.
-  //
-  // The dummy values can be uninitialized. cusparseSpGEMM_compute does
-  // not remove numerical zeros from the sparsity pattern.
-  void *dummyValues;
-  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(
-      &dummyValues,
-      sizeof(scalar_type) * std::max(entriesA.extent(0), entriesB.extent(0))));
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
+        &h->descr_A, m, n, entriesA.extent(0), (void *)row_mapA.data(),
+        (void *)entriesA.data(), dummyValues_AB, CUSPARSE_INDEX_32I,
+        CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
 
-  KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
-      &h->descr_A, m, n, entriesA.extent(0), (void *)row_mapA.data(),
-      (void *)entriesA.data(), dummyValues, CUSPARSE_INDEX_32I,
-      CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
+        &h->descr_B, n, k, entriesB.extent(0), (void *)row_mapB.data(),
+        (void *)entriesB.data(), dummyValues_AB, CUSPARSE_INDEX_32I,
+        CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
 
-  KOKKOS_CUSPARSE_SAFE_CALL(cusparseCreateCsr(
-      &h->descr_B, n, k, entriesB.extent(0), (void *)row_mapB.data(),
-      (void *)entriesB.data(), dummyValues, CUSPARSE_INDEX_32I,
-      CUSPARSE_INDEX_32I, CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
+    KOKKOS_CUSPARSE_SAFE_CALL(
+        cusparseCreateCsr(&h->descr_C, m, k, 0, row_mapC.data(), nullptr,
+                          nullptr, CUSPARSE_INDEX_32I, CUSPARSE_INDEX_32I,
+                          CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
 
-  KOKKOS_CUSPARSE_SAFE_CALL(
-      cusparseCreateCsr(&h->descr_C, m, k, 0, row_mapC.data(), nullptr, nullptr,
-                        CUSPARSE_INDEX_32I, CUSPARSE_INDEX_32I,
-                        CUSPARSE_INDEX_BASE_ZERO, h->scalarType));
+    //----------------------------------------------------------------------
+    // query workEstimation buffer size, allocate, then call again with buffer.
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_workEstimation(
+        h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B,
+        &beta, h->descr_C, h->scalarType, h->alg, h->spgemmDescr,
+        &h->bufferSize3, nullptr));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        cudaMalloc((void **)&h->buffer3, h->bufferSize3));
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_workEstimation(
+        h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B,
+        &beta, h->descr_C, h->scalarType, h->alg, h->spgemmDescr,
+        &h->bufferSize3, h->buffer3));
 
-  //----------------------------------------------------------------------
-  // query workEstimation buffer size, allocate, then call again with buffer.
-  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_workEstimation(
-      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
-      h->descr_C, h->scalarType, h->alg, h->spgemmDescr, &h->bufferSize3,
-      nullptr));
-  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc((void **)&h->buffer3, h->bufferSize3));
-  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_workEstimation(
-      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
-      h->descr_C, h->scalarType, h->alg, h->spgemmDescr, &h->bufferSize3,
-      h->buffer3));
-  cudaFree(h->buffer3);
-  h->buffer3 = nullptr;
+    //----------------------------------------------------------------------
+    // query compute buffer size, allocate, then call again with buffer.
 
-  //----------------------------------------------------------------------
-  // query compute buffer size, allocate, then call again with buffer.
-
-  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_compute(
-      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
-      h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT, h->spgemmDescr,
-      &h->bufferSize4, nullptr));
-  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc((void **)&h->buffer4, h->bufferSize4));
-  KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_compute(
-      h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B, &beta,
-      h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT, h->spgemmDescr,
-      &h->bufferSize4, h->buffer4));
-  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFree(dummyValues));
-
-  int64_t C_nrow, C_ncol, C_nnz;
-  KOKKOS_CUSPARSE_SAFE_CALL(
-      cusparseSpMatGetSize(h->descr_C, &C_nrow, &C_ncol, &C_nnz));
-  if (C_nnz > std::numeric_limits<int>::max()) {
-    throw std::runtime_error("nnz of C overflowed over 32-bit int\n");
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_compute(
+        h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B,
+        &beta, h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT,
+        h->spgemmDescr, &h->bufferSize4, nullptr));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        cudaMalloc((void **)&h->buffer4, h->bufferSize4));
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_compute(
+        h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B,
+        &beta, h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT,
+        h->spgemmDescr, &h->bufferSize4, h->buffer4));
+    int64_t C_nrow, C_ncol, C_nnz;
+    KOKKOS_CUSPARSE_SAFE_CALL(
+        cusparseSpMatGetSize(h->descr_C, &C_nrow, &C_ncol, &C_nnz));
+    if (C_nnz > std::numeric_limits<int>::max()) {
+      throw std::runtime_error("nnz of C overflowed over 32-bit int\n");
+    }
+    handle->set_c_nnz(C_nnz);
+    handle->set_call_symbolic();
+    firstSymbolicCall = true;
   }
-  handle->set_c_nnz(C_nnz);
-  handle->set_call_symbolic();
-  handle->set_computed_rowptrs();
+
+  if (computeRowptrs && !handle->are_rowptrs_computed()) {
+    auto h     = handle->get_cusparse_spgemm_handle();
+    auto C_nnz = handle->get_c_nnz();
+    if (!firstSymbolicCall) {
+      // This is not the first call to symbolic, so dummyValues_AB was not
+      // allocated above. But, descr_A and descr_B will have been saved in the
+      // handle, so we can reuse those.
+      KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(
+          &dummyValues_AB, sizeof(scalar_type) * std::max(entriesA.extent(0),
+                                                          entriesB.extent(0))));
+      KOKKOS_CUSPARSE_SAFE_CALL(
+          cusparseCsrSetPointers(h->descr_A, (void *)row_mapA.data(),
+                                 (void *)entriesA.data(), dummyValues_AB));
+      KOKKOS_CUSPARSE_SAFE_CALL(
+          cusparseCsrSetPointers(h->descr_B, (void *)row_mapB.data(),
+                                 (void *)entriesB.data(), dummyValues_AB));
+    }
+    void *dummyEntries_C, *dummyValues_C;
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        cudaMalloc(&dummyEntries_C, sizeof(ordinal_type) * C_nnz));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(
+        cudaMalloc(&dummyValues_C, sizeof(scalar_type) * C_nnz));
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseCsrSetPointers(
+        h->descr_C, (void *)row_mapC.data(), dummyEntries_C, dummyValues_C));
+
+    KOKKOS_CUSPARSE_SAFE_CALL(cusparseSpGEMM_copy(
+        h->cusparseHandle, h->opA, h->opB, &alpha, h->descr_A, h->descr_B,
+        &beta, h->descr_C, h->scalarType, CUSPARSE_SPGEMM_DEFAULT,
+        h->spgemmDescr));
+
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFree(dummyValues_C));
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFree(dummyEntries_C));
+    handle->set_computed_rowptrs();
+  }
+  KOKKOS_IMPL_CUDA_SAFE_CALL(cudaFree(dummyValues_AB));
 }
 
 #else

--- a/sparse/unit_test/Test_Sparse_spgemm.hpp
+++ b/sparse/unit_test/Test_Sparse_spgemm.hpp
@@ -407,6 +407,70 @@ void test_spgemm(lno_t m, lno_t k, lno_t n, size_type nnz, lno_t bandwidth,
 
 template <typename scalar_t, typename lno_t, typename size_type,
           typename device>
+void test_spgemm_symbolic(bool callSymbolicFirst, bool testEmpty) {
+  using crsMat_t       = CrsMatrix<scalar_t, lno_t, device, void, size_type>;
+  using graph_t        = typename crsMat_t::StaticCrsGraphType;
+  using values_t       = typename crsMat_t::values_type;
+  using entries_t      = typename graph_t::entries_type;
+  using rowmap_t       = typename graph_t::row_map_type::non_const_type;
+  using const_rowmap_t = typename graph_t::row_map_type;
+  using KernelHandle   = KokkosKernels::Experimental::KokkosKernelsHandle<
+      size_type, lno_t, scalar_t, typename device::execution_space,
+      typename device::memory_space, typename device::memory_space>;
+  // A is m*n, B is n*k, C is m*k
+  int m = 100;
+  int n = 300;
+  int k = 200;
+  crsMat_t A, B;
+  // Target 1000 total nonzeros in both A and B.
+  if (testEmpty) {
+    // Create A,B with the same dimensions, but zero entries
+    values_t emptyValues;
+    entries_t emptyEntries;
+    // Initialize these to 0
+    rowmap_t A_rowmap("A rowmap", m + 1);
+    rowmap_t B_rowmap("B rowmap", n + 1);
+    A = crsMat_t("A", m, n, 0, emptyValues, A_rowmap, emptyEntries);
+    B = crsMat_t("B", n, k, 0, emptyValues, B_rowmap, emptyEntries);
+  } else {
+    size_type nnz = 1000;
+    A   = KokkosSparse::Impl::kk_generate_sparse_matrix<crsMat_t>(m, n, nnz, 10,
+                                                                50);
+    nnz = 1000;
+    B   = KokkosSparse::Impl::kk_generate_sparse_matrix<crsMat_t>(n, k, nnz, 10,
+                                                                50);
+    KokkosSparse::sort_crs_matrix(A);
+    KokkosSparse::sort_crs_matrix(B);
+  }
+  // Call reference impl to get complete product
+  crsMat_t C_reference;
+  Test::run_spgemm<crsMat_t, device>(A, B, SPGEMM_DEBUG, C_reference, false);
+  // Now call just symbolic, and specifically request that rowptrs be populated
+  // Make sure this never depends on C_rowmap being initialized
+  rowmap_t C_rowmap(Kokkos::view_alloc(Kokkos::WithoutInitializing, "rowmapC"),
+                    m + 1);
+  Kokkos::deep_copy(C_rowmap, size_type(123));
+  KernelHandle kh;
+  kh.create_spgemm_handle();
+  if (callSymbolicFirst) {
+    KokkosSparse::Experimental::spgemm_symbolic(
+        &kh, m, n, k, A.graph.row_map, A.graph.entries, false, B.graph.row_map,
+        B.graph.entries, false, C_rowmap);
+  }
+  KokkosSparse::Experimental::spgemm_symbolic(
+      &kh, m, n, k, A.graph.row_map, A.graph.entries, false, B.graph.row_map,
+      B.graph.entries, false, C_rowmap, true);
+  kh.destroy_spgemm_handle();
+  bool isCorrect = KokkosKernels::Impl::kk_is_identical_view<
+      const_rowmap_t, const_rowmap_t, size_type,
+      typename device::execution_space>(C_rowmap, C_reference.graph.row_map, 0);
+  EXPECT_TRUE(isCorrect)
+      << " spgemm_symbolic produced incorrect rowptrs - callSymbolicFirst = "
+      << callSymbolicFirst << ", empty A/B = " << testEmpty;
+}
+
+template <typename scalar_t, typename lno_t, typename size_type,
+          typename device>
 void test_issue402() {
 #if defined(KOKKOSKERNELS_ENABLE_TPL_ARMPL)
   {
@@ -502,6 +566,10 @@ void test_issue402() {
     test_spgemm<SCALAR, ORDINAL, OFFSET, DEVICE>(10, 10, 0, 0, 10, 10, true);  \
     test_spgemm<SCALAR, ORDINAL, OFFSET, DEVICE>(10, 10, 10, 0, 0, 0, false);  \
     test_spgemm<SCALAR, ORDINAL, OFFSET, DEVICE>(10, 10, 10, 0, 0, 0, true);   \
+    test_spgemm_symbolic<SCALAR, ORDINAL, OFFSET, DEVICE>(true, true);         \
+    test_spgemm_symbolic<SCALAR, ORDINAL, OFFSET, DEVICE>(false, true);        \
+    test_spgemm_symbolic<SCALAR, ORDINAL, OFFSET, DEVICE>(true, false);        \
+    test_spgemm_symbolic<SCALAR, ORDINAL, OFFSET, DEVICE>(false, false);       \
     test_issue402<SCALAR, ORDINAL, OFFSET, DEVICE>();                          \
   }
 

